### PR TITLE
feat(ME-785): connector auth support

### PR DESCRIPTION
**Pull request by btoonk**

### CODE REVIEW

File: cmd/client/db/dbeaver.go [hunk 1]
```diff
@@ -117,6 +140,13 @@ var dbeaverCmd = &cobra.Command{
 		default:
 			err = client.ExecCommand("dbeaver", "-con", conn)
 		}
+
+		if info.ConnectorAuthenticationEnabled {
+			ch := make(chan os.Signal, 1)
+			signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+			<-ch
```
Comment (pedroandrade) [inline cmd/client/db/dbeaver.go:147]: is this to keep the process run?

File: cmd/client/tls/tls.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+					}
+
+					log.Print("Connection established from ", lcon.RemoteAddr())
+					tcp_con_handle(conn, lcon, lcon)
+				}()
+			}
+		} else {
+			var conn net.Conn
+			if info.ConnectorAuthenticationEnabled {
```
Comment (pedroandrade) [inline cmd/client/tls/tls.go:103]: I think this is the same code as line 82 to 95, I guess, should we extract it to a function? something like `establishConnection() error`

File: cmd/client/tls/tls.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+				if err != nil {
+					log.Fatalf("failed to connect: %v", err.Error())
+				}
+			} else {
+				conn, err = tls.Dial("tcp", fmt.Sprintf("%s:%d", hostname, info.Port), &tlsConfig)
+				if err != nil {
+					log.Fatalf("failed to connect: %v", err.Error())
+				}
+			}
+			tcp_con_handle(conn, os.Stdin, os.Stdout)
```
Comment (pedroandrade) [inline cmd/client/tls/tls.go:114]: camel case function name to keep the standards :D

File: cmd/client/tls/tls.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	chan_to_remote := stream_copy(in, con)
+
+	select {
+	case <-chan_to_stdout:
+	case <-chan_to_remote:
+	}
+}
+
+// Performs copy operation between streams: os and tcp streams
+func stream_copy(src io.Reader, dst io.Writer) <-chan int {
```
Comment (pedroandrade) [inline cmd/client/tls/tls.go:132]: camel case function name to keep the standards :D

File: cmd/client/tls/tls.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	buf := make([]byte, 1024)
+	sync_channel := make(chan int)
```
Comment (pedroandrade) [inline cmd/client/tls/tls.go:134]: camelCase var names

File: internal/ssh/server.go [hunk 6]
```diff
@@ -44,20 +50,17 @@ func newServer(ca string) *ssh.Server {
...
-			} else {
-				cmd.Args = []string{shell, "-c", s.RawCommand()}
-			}
-		} else {
-			cmd.Path = shell
-			if runtime.GOOS != "windows" {
-				cmd.Args = []string{"-" + filepath.Base(shell)}
-			}
-		}
+		// if len(s.Command()) > 0 {
```
Comment (pedroandrade) [inline internal/ssh/server.go:53]: we don't need this commented code, right?

File: internal/ssh/server.go [hunk 7]
```diff
@@ -69,11 +72,39 @@ func newServer(ca string) *ssh.Server {
 		cmd.Dir = user.HomeDir
 
 		execCmd(s, cmd, uid, gid)
-
 	})
 
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		log.Printf("could not generate rsa key: %s", err)
+		return nil
```
Comment (pedroandrade) [inline internal/ssh/server.go:80]: just curious, shouldn't we return the error in this function?

File: internal/ssh/server_unix.go [hunk 8]
```diff
@@ -19,44 +19,85 @@ import (
...
-		cmd.Wait()
-	} else {
-		cmd.SysProcAttr = &syscall.SysProcAttr{
-			Credential: &syscall.Credential{
-				Uid:         uint32(uid),
-				Gid:         uint32(gid),
-				NoSetGroups: true,
-			},
-			Setsid: true,
+		go func() {
```
Comment (pedroandrade) [inline internal/ssh/server_unix.go:78]: not for now, at some point we should replace these goroutines calls with ```https://pkg.go.dev/golang.org/x/sync/errgroup``` structure

Comment (waltzofpearls) [reply internal/ssh/server_unix.go:78]: Agreed, `errgroup` understands `context.Context`, and coordinate goroutines with wait group.

**Review by waltzofpearls — APPROVED** [review]
Looking great overall. I found 2 small things and left comments.

Like Pedro said in one comment, there are some duplicate code that was probably copied from one file to another to support connector auth in different commands. It would be great if those could be abstracted into reusable functions.

File: cmd/client/db/dbeaver.go (hunk 1 above)
Comment (btoonk) [reply cmd/client/db/dbeaver.go:147]: Yes, it's a channel that is waiting for a signal (ctr^c)

File: cmd/client/db/mysql.go [hunk 9]
```diff
@@ -64,12 +65,27 @@ var mysqlCmd = &cobra.Command{
 		defer persistPreference()
 		client.OnInterruptDo(persistPreference)
 
+		if info.ConnectorAuthenticationEnabled {
```
[the remark this answers is not in the document]
Comment (btoonk) [reply cmd/client/db/mysql.go:68]: Do you have any suggestions on how to improve this?

File: cmd/client/tls/tls.go (hunk 2 above)
Comment (btoonk) [reply cmd/client/tls/tls.go]: done

File: cmd/client/tls/tls.go (hunk 3 above)
Comment (btoonk) [reply cmd/client/tls/tls.go]: done

File: cmd/client/tls/tls.go (hunk 4 above)
Comment (btoonk) [reply cmd/client/tls/tls.go]: done

File: internal/ssh/server.go (hunk 6 above)
Comment (btoonk) [reply internal/ssh/server.go]: no, removed

File: internal/ssh/server.go (hunk 7 above)
Comment (btoonk) [reply internal/ssh/server.go]: Changed it to a fatal error.

File: internal/ssh/server_unix.go (hunk 8 above)
Comment (btoonk) [reply internal/ssh/server_unix.go:78]: ack

File: cmd/client/db/dbeaver.go [hunk 10]
```diff
@@ -50,22 +53,41 @@ var dbeaverCmd = &cobra.Command{
...
+
+		if info.ConnectorAuthenticationEnabled {
+			certificate := tls.Certificate{
+				Certificate: [][]byte{info.Certficate.Raw},
+				PrivateKey:  info.PrivateKey,
+			}
+
+			info.Port, err = client.StartConnectorAuthListener(fmt.Sprintf("%s:%d", hostname, info.Port), certificate, 0)
+			if err != nil {
+				fmt.Println("ERROR: could not setup listener:", err)
```
[the remark this answers is not in the document]
Comment (btoonk) [reply cmd/client/db/dbeaver.go]: done

File: cmd/client/tls/tls.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+			if err != nil {
+				return err
+			}
+			hostname = pickedHost.Hostname()
+		}
+
+		//Check for  hostname checking in *.border0-dummy
+		// This may be used by ssh users
+		// if so strip that
+		substr := "(.*).border0-dummy$"
```
[the remark this answers is not in the document]
Comment (btoonk) [reply cmd/client/tls/tls.go]: done

**Review by pedroandrade — APPROVED** [review]

**Pull request merged**